### PR TITLE
Refactor and add JSDoc support

### LINKED ISSUE #1: feat: check TSDoc
There are many code repositories for grammY and they all contain docs. While most of them don't have many links, it would still be comforting to know that the links that are contained are correct.

### PR DESCRIPTION
**Pull request by dcdunkan**
Re-wrote the whole thing once again. Fixed few mistakes, and made new mistakes.

- Pretty issue reports and neat summary.
- Ability to check for broken links in a JS/TS module.
- Groupable links (plugin?). Some types of links needs to be taken care of in a special way.
- CLIs have been improved and split into a different file.
- WAY better redirection validator.

Closes #1

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,16 +1,53 @@
 # Link Checker
 
-A tool for going through
-[grammY documentation source](https://github.com/grammyjs/website) and reporting
-broken links.
+A tool for going through grammY documentation source and other repositories' JSDocs, and reporting broken links and missing
```
Comment (rojvv) [inline README.md]:
```suggestion
A tool for going through grammY documentation source and other repositories’ JSDocs, and reporting broken links and missing
```

File: README.md [hunk 2]
```diff
@@ -1,16 +1,53 @@
...
+anchors.
 
----
+## Checking Documentation
 
-To use the tool locally, clone the
-[grammY website repository](https://github.com/grammyjs/website) and go to the
-`site/docs/` directory and run the following:
+To check for broken links in [grammY's documentation source](https://github.com/grammyjs/website), clone the website repository
```
Comment (rojvv) [inline README.md]:
```suggestion
To check for broken links in the [grammY documentation](https://github.com/grammyjs/website), clone the website repository
```

File: README.md [hunk 3]
```diff
@@ -1,16 +1,53 @@
...
-  https://raw.githubusercontent.com/grammyjs/link-checker/main/main.ts
+$ deno run --allow-env --allow-net --allow-read \
+  https://raw.githubusercontent.com/grammyjs/link-checker/main/website_cli.ts [ROOT DIRECTORY]
 ```
+
+> Assumes the current directory as the ROOT DIRECTORY if none specified.
+
+Supported arguments:
+
+- `--clean-url`: Defaults to `false`. Vitepress's clean URL config option makes sure that the references to other local files
```
Comment (rojvv) [inline README.md]:
```suggestion
- `--clean-url`: Defaults to `false`. VitePress’ clean URL config option makes sure that the references to other local files
```

File: README.md [hunk 4]
```diff
@@ -1,16 +1,53 @@
...
+  doesn't end with an extension. Specify this option if the configuration used in the website repository has this option enabled.
```
Comment (rojvv) [inline README.md]:
```suggestion
  doesn’t end with an extension. Specify this option if the configuration used in the website repository has this option enabled.
```

File: README.md [hunk 5]
```diff
@@ -1,16 +1,53 @@
...
+**TLDR;** Set the environment variable `GITHUB_TOKEN` if documents contains a lot of links to GitHub markup (README, for example)
+files. A fine-grained token with the default set of permissions will do.
+
+When you run the CLIs as mentioned you should get a gentle reminder about an environment variable: `GITHUB_TOKEN`. Here's why:
+There might be links like https://github.com/OWNER/REPO/tree/dir#anchor or https://github.com/OWNER/REPO/blob/file.rst#anchor in
+the documentation. These types of links are assumed to be pointing to a GitHub repository's directory README or a
+[supported markup](https://github.com/github/markup/blob/master/README.md#markups) file, because there is an anchor. When normally
+fetched, their responses does not contain the actual content of that file, because the content is lazily loaded. So for checking
+if the anchor is present, we fetch the content using the GitHub Content REST API. And that is why we need a GitHub access token.
+If you only have a few such links, then you most likely don't need the token.
```
Comment (rojvv) [inline README.md]:
```suggestion
If you only have a few such links, then you most likely don’t need the token.
```

File: README.md [hunk 6]
```diff
@@ -1,16 +1,53 @@
...
+  reference other local files.
+
+#### GitHub files
+
+**TLDR;** Set the environment variable `GITHUB_TOKEN` if documents contains a lot of links to GitHub markup (README, for example)
+files. A fine-grained token with the default set of permissions will do.
+
+When you run the CLIs as mentioned you should get a gentle reminder about an environment variable: `GITHUB_TOKEN`. Here's why:
+There might be links like https://github.com/OWNER/REPO/tree/dir#anchor or https://github.com/OWNER/REPO/blob/file.rst#anchor in
+the documentation. These types of links are assumed to be pointing to a GitHub repository's directory README or a
```
Comment (rojvv) [inline README.md]:
```suggestion
the documentation. These types of links are assumed to be pointing to a GitHub repository’s directory README or a
```

File: README.md [hunk 7]
```diff
@@ -1,16 +1,53 @@
...
+- `--index-file`: Defaults to `README.md`. Index file of a directory.
+- `--allow-ext-html`: Defaults to `false`. Not needed if `--clean-url` is specified. Allows `.html` extension to be used to
+  reference other local files.
+
+#### GitHub files
+
+**TLDR;** Set the environment variable `GITHUB_TOKEN` if documents contains a lot of links to GitHub markup (README, for example)
+files. A fine-grained token with the default set of permissions will do.
+
+When you run the CLIs as mentioned you should get a gentle reminder about an environment variable: `GITHUB_TOKEN`. Here's why:
```
Comment (rojvv) [inline README.md]:
```suggestion
When you run the CLIs as mentioned you should get a gentle reminder about an environment variable: `GITHUB_TOKEN`. Here’s why:
```

**Comment by quadratz** [comment]
Please correct me if I'm mistaken, but if the link checker requires a GitHub token, it is designed to be used specifically with GitHub Action, correct? Unless we are comfortable hard coding the token in our code, users will need to utilize their own GitHub token and run it on their individual PCs.

```console
➜  deno run -r --allow-env --allow-net --allow-read website_cli.ts ./                                          

┃ Gentle reminder: It is recommended to set the GITHUB_TOKEN environment variable
┃ if there are GitHub repository file links in the documents.

reading README.md
It's a renderable Github file with an anchor. Resolving using Github API...
error: Uncaught Error: Unauthorized
  if (!response.ok) throw new Error(response.statusText);
                          ^
```

**Comment by dcdunkan** [comment]
Thanks for catching that. That was a mistake, that part is wrong. Will fix that. I realized that I actually never tested it without a token after I completed it.

Link checker doesn't "require" the token unless there are a lot of such links. Unauthorized requests have 60 req/h limit. And no, it is not designed specifically to use with GitHub Actions. But for a repository like our website's, it is reasonable to assume that we are gonna need a token, because we can't simply predict the number of such links (even tho we currently only have 1 link, and only two request are made, one for branches and the other for the rendered file content).

Here it was supposed to be leave the reminder and fetch it without the token. But I got lost.

**Comment by dcdunkan** [comment]
Can you try it now? (without a token)

**Comment by dcdunkan** [comment]
> ```
> ➜  deno run -r --allow-env --allow-net --allow-read website_cli.ts ./
> ```

You're gonna need `--clean-url` :D

**Comment by quadratz** [comment]
Yup. It's working now. Let me do some testing

**Review by quadratz — COMMENTED** [review]
There should be at least 11 errors, not 5.

```console
➜  deno run -r --allow-env --allow-net --allow-read website_cli.ts --clean-url test/clean-url/

┃ Gentle reminder: It is recommended to set the GITHUB_TOKEN environment variable
┃ if there are GitHub repository file links in the documents.

reading test\clean-url\bar.md
reading test\clean-url\foo.md
reading test\clean-url\id\bar.md
reading test\clean-url\id\foo.md
reading test\clean-url\id\guide\bar.md
reading test\clean-url\id\guide\foo.md
reading test\clean-url\id\guide\README.md
reading test\clean-url\id\README.md
reading test\clean-url\invalid\bar.md
reading test\clean-url\invalid\es\foo.md
reading test\clean-url\invalid\es\guide\foo.md
reading test\clean-url\invalid\foo.md
reading test\clean-url\README.md
Found 5 issues in the documentation:

┌────────────────────────┬───┐
│          Missing files │ 5 │
├────────────────────────┼───┤
│                  Total │ 5 │
└────────────────────────┴───┘

test\clean-url\invalid\bar.md (4)
— The linked file ../id/bazzz.md does not exist.
— The linked file ./lost-file.md does not exist.
— The linked file ./es/lost-file.md does not exist.
— The linked file ./es/guide/lost-file.md does not exist.

test\clean-url\invalid\foo.md (1)
— The linked file ./id/bar.md does not exist.
```

**Comment by dcdunkan** [comment]
> There should be at least 11 errors, not 5.

It'd be easier to fix if could you post all the expected ones?

**Comment by quadratz** [comment]
@dcdunkan Just one more left. This url https://deno.land/x/grammy@v0.00.0 should return 404

**Comment by quadratz** [comment]
`200 OK`? Really?! 🤯 

image

image

**Comment by dcdunkan** [comment]
Yeah, I also checked. Consider opening an issue at denoland/website_feedback. This shouldn't be adjusted at our side.

**Review by quadratz — APPROVED** [review]
LGTM. The link-checker passed all of my tests, and I also tested it on the grammy/website repository. Thank you for your hard work.

**Comment by quadratz** [comment]
Should we remove the test folder, @dcdunkan?

**Comment by dcdunkan** [comment]
> LGTM. The link-checker passed all of my tests, and I also tested it on the grammy/website repository. Thank you for your hard work.

Awesome, thank you so much for testing and catching my mistakes.

**Comment by dcdunkan** [comment]
> Should we remove the test folder, @dcdunkan?

Yeah. Writing tests for this was in my head when I started, then I got lazy. But now that you have already did this, I'd like to write some actual tests. I'll take care of that now. Might need another review for that too :)

**Comment by quadratz** [comment]
Aye aye capt!

**Comment by dcdunkan** [comment]
@CikiMomogi Please revert 35887f1. That's not a valid redirection.

They have plans about merging deno.com and deno.land. They first redirected certain links from land to com and then they didn't change anything for some time, now this too, com to land. Since land's source code is archived now and com has always been closed source, any changes could take place, that we don't know anything about. @KnorpelSenf and I talked about this and have decided to do nothing about them until they settle for enough time. That's why this is a invalid redirection.

When I fixed the links in website, there was this redirection:
https://deno.land to https://deno.com/runtime. And for a while they stayed like that, and they changed it back, no redirections. Then again, they now redirects deno.com/runtime to deno.land lol. So, skip it.

Reference about their merging plans: https://github.com/denoland/website_feedback/issues/34

**Comment by quadratz** [comment]
I suggest keeping it to deno.land for the time being since we're not entirely sure when the link will be fixed.

**Comment by dcdunkan** [comment]
I don't agree to that, sorry. If some other intended deno.com link does a valid redirection to deno.com itself, then that would be an invalid redirection since we're expecting deno.land. So, I'd rather prefer fixing these issues in the website repository; or as discussed, wait a while and then do something about it.

**Comment by quadratz** [comment]
@dcdunkan, is this PQ good to merge? I'm pretty sure it'll fix the anchor-check error in the current link-checker.

**Comment by KnorpelSenf** [comment]
I think the whole deno.com/deno.land thing has settled now, too, so we can finish this.

**Comment by dcdunkan** [comment]
@quadratz @KnorpelSenf Sorry about the late response. Nothing important changed in the last 4 commits. So, I'll just merge it now. I hope @quadratz can send the update PQ in /website? Thanks for reviewing and catching my mistakes!

(I was lazy again, so I didn't add any tests)

**Pull request merged**